src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs

```diff
@@ -791,18 +768,41 @@ protected TNode AddError<TNode>(TNode nodeOrToken, ErrorCode code, params object
             // token, and we will want to place the diagnostic on it.  Otherwise, we want to place it on the true 'next
             // token' the parser is currently pointing at.
 
-            if (!nodeOrToken.ContainsSkippedText)
+            if (!missingNodeOrToken.ContainsSkippedText)
             {
                 // Simple case this node/token does not contain any skipped text.  Place the diagnostic at the start of
                 // the token that follows.
-                var (offset, width) = this.GetDiagnosticSpanForMissingToken();
-                return WithAdditionalDiagnostics(nodeOrToken, MakeError(nodeOrToken.FullWidth + offset, width, code, args));
+                return getOffsetAndWidthBasedOnPriorAndNextTokens();
             }
             else
             {
                 // Complex case.  This node or token contains skipped text.  Place the diagnostic on the skipped text.
-                var (offset, width) = getOffsetAndWidthOfSkippedToken();
-                return WithAdditionalDiagnostics(nodeOrToken, MakeError(offset, width, code, args));
+                return getOffsetAndWidthOfSkippedToken();
+            }
+
+            (int offset, int width) getOffsetAndWidthBasedOnPriorAndNextTokens()
+            {
+                // If the previous token has a trailing EndOfLineTrivia, the missing token diagnostic position is moved
+                // to the end of line containing the previous token and its width is set to zero. Otherwise we squiggle
+                // the token following the missing token (the token we're currently pointing at).
+
+                var trivia = _prevTokenTrailingTrivia;
+                var triviaList = new SyntaxList<CSharpSyntaxNode>(_prevTokenTrailingTrivia);
+                if (triviaList.Any((int)SyntaxKind.EndOfLineTrivia))
+                {
+                    // Offsets is relative to the start (not full start) of the missing node/token. So to move before
+                    // the preceding trivia, we have to subtract the leading trivia on the node, and then the trailing
+                    // trivia on the previous token.
+                    return (offset: -missingNodeOrToken.GetLeadingTriviaWidth() - trivia.FullWidth, width: 0);
+                }
+                else
+                {
+                    // Offsets is relative to the start (not full start) of the missing node/token. So to move to the location
+                    // of the next token we have to move past missing token's content, and it's trailing trivia, and the leading
+                    // trivia of the next token.
+                    var token = this.CurrentToken;
+                    return (missingNodeOrToken.Width + missingNodeOrToken.GetTrailingTriviaWidth() + token.GetLeadingTriviaWidth(), token.Width);
```

Comment:
this is similarly updated.  to accurately move to the start of the next token, we need to add the current width, trailing trivia width *and* next-token-leading-trivia-width.